backend/cn/lib/eval.ml

```diff
@@ -0,0 +1,606 @@
+module BT = BaseTypes
+module IT = IndexTerms
+
+let check_bits_bt (sgn1, sz1) (sgn2, sz2) here : (unit, string) result =
+  if Stdlib.( = ) sgn1 sgn2 then
+    Result.error ("Mismatched signs at " ^ Locations.simple_location here)
+  else if sz1 <> sz2 then
+    Result.error ("Mismatched sizes at " ^ Locations.simple_location here)
+  else
+    Ok ()
+
+
+let handle_bounds (sgn, sz) (n : Z.t) : (Z.t, string) result =
+  let min, max = BT.bits_range (sgn, sz) in
+  match sgn with
+  | Unsigned ->
+    Ok
+      (if Z.gt n max then
+         Z.rem n (Z.add max Z.one)
+       else if Z.lt n Z.zero then
+         Z.add n max
+       else
+         n)
```

Comment:
If `n` can go arbitrarily far out of the (min,max) range, then this has to be recursive. Maybe there's a stronger invariant that avoids that, however.